Reconcile neighbors after the vm live migration on default cluster network.

qinqon opened the pull request:
## Description

After VM live migration to a different node, neighbor cache entries (ARP/NDP)
become stale and must be updated to ensure proper traffic routing. This PR adds
neighbor reconciliation logic that sends GARPs and unsolicited NAs to update
ARP/NDP caches with the arp_proxy MAC address when VMs migrate between nodes
with different subnet ownership.

### Gateway IP neighbor fix

After PR#5773 changed DHCP to advertise the actual subnet gateway IP (e.g.,
10.244.0.1) instead of the link-local 169.254.1.1 address, VMs resolve the
gateway to the node's real LRP MAC on the subnet-owning switch (the LRP's ARP
responder has higher priority than arp_proxy). After migration, this LRP MAC
doesn't exist on the destination switch, so we must also GARP the gateway IP to
update the VM's neighbor entry to the arp_proxy MAC. This issue was masked in
e2e tests by PR#4755 which changed the connectivity test checks.

References:
- Bug introduced by: https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5773
- Masked in tests by: https://github.com/ovn-kubernetes/ovn-kubernetes/pull/4755

### Reconciliation details

The reconciliation handles two scenarios:
- **At migration target (local zone):** notify the VM that subnet peers and the
  gateway IP should use the arp_proxy MAC via GARPs/unsolicited NAs.
- **At subnet-owning node (remote zone):** delete stale source pod LSPs to
  prevent wrong ARP responses, and notify remaining pods to use the arp_proxy
  MAC for the migrated VM's IP.

### E2E test refactor

This also refactors the kubevirt e2e tests related to default pod network
migration so they use iperf3 for north-south and east-west connectivity,
ensuring downtime measurement at east-west and removing the custom TCP server
that was previously used.

Fixes #3986

## Checks
- [x] My code requires changes to the documentation
- [x] if so, I have updated the documentation as required
- [x] My code requires tests
- [x] if so, I have added and/or updated the tests as required
- [ ] All the tests have passed in the CI

## How to verify it
It includes e2e tests

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **Documentation**
  * Expanded live-migration guide with detailed neighbor advertising workflow, GARP handling, and multi-zone neighbor table updates.

* **New Features**
  * Enhanced live migration support with zone-aware pod routing and neighbor reconciliation for improved network consistency across migrations.

* **Tests**
  * Updated live migration tests to validate multiple migration modes and dynamic network configurations.

* **Chores**
  * Removed legacy test infrastructure components no longer in use.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

qinqon committed:
kv, design: Advertise neighbors

This change add the design to advertise neighbors after kubevirt's VM
live migration to improve connectivity

Signed-off-by: Enrique Llorente <ellorent@redhat.com>

qinqon committed:
kv, e2e: Refactor cluster default network to use iperf3

Now the kubevirt cluser default network tests run exactly like the UDNs
one this change remove all the custom server and we use also the same VM
image.

Signed-off-by: Enrique Llorente <ellorent@redhat.com>

qinqon commented:
/retest

qinqon commented:
/retest

kyrtapz commented on docs/features/live-migration.md:
````diff
@@ -428,6 +429,80 @@ The point to point routing cleanup (remove of static routes and policies) will b
+```text
+    ┌────────────────────────┐ ┌────────────────────────┐
+    │logical switch node2    │ │logic switch node1      │ 
+    │   (10.244.1.0/24)      │ │   (10.244.0.0/24)      │
````
nit but makes sense.

kyrtapz commented on go-controller/pkg/kubevirt/router.go:
```diff
@@ -166,33 +179,73 @@ func EnsureLocalZonePodAddressesToNodeRoute(watchFactory *factory.WatchFactory,
-//     current node transit switch port.
-func EnsureRemoteZonePodAddressesToNodeRoute(watchFactory *factory.WatchFactory, nbClient libovsdbclient.Client, pod *corev1.Pod, nadName string) error {
+func reconcileClusterDefaultNetworkLocalZoneNeighbours(watchFactory *factory.WatchFactory, pod *corev1.Pod, podAnnotation *util.PodAnnotation, nadName string) error {
+	// TODO: Implement neighbours update for non IC
```
Is this something you are planning to do? Is there any upstream issue tracking it that we can link?

qinqon replied:
We are planning to not support this for non IC, I don't think there are any live migration for default pod network non IC users.

kyrtapz commented on go-controller/pkg/kubevirt/pod.go:
```diff
@@ -683,3 +711,63 @@ func newRouterAdvertisementWithPrefixInfos(ip net.IP, destinationMAC net.Hardwar
+		return nil, err
+	}
+	if nodeOwningSubnet == nil {
+		return nil, nil
```
Shouldn't this be an error?

qinqon replied:
For example if we migrate a VM, and then that node goes down this will be the case and since we don't have nothing to GARP for we just return empty here.

kyrtapz commented on go-controller/pkg/kubevirt/pod.go:
```diff
@@ -683,3 +711,63 @@ func newRouterAdvertisementWithPrefixInfos(ip net.IP, destinationMAC net.Hardwar
+			continue
+		}
+
+		//FIXME: Is comparing MAC enough to discard pod's from same VM
```
Doesn't kubevirt store the VM identifier in a label or something?

kyrtapz commented on go-controller/pkg/kubevirt/pod.go:
```diff
@@ -683,3 +711,63 @@ func newRouterAdvertisementWithPrefixInfos(ip net.IP, destinationMAC net.Hardwar
 		PrefixInfos:    prefixInfos,
 	}
 }
+func findRunningPodsIPsFromPodSubnet(watchFactory *factory.WatchFactory, podAnnotation *util.PodAnnotation, nadName string) ([]*net.IPNet, error) {
```
I think this function could use a comment on what it does.

kyrtapz commented on go-controller/pkg/kubevirt/router.go:
```diff
@@ -166,33 +179,73 @@ func EnsureLocalZonePodAddressesToNodeRoute(watchFactory *factory.WatchFactory,
+	// Now that the VM is running at a node not owning his subnet, using pod's mac
+	// will not work since it's a different switch so overwrite those neighbors with
+	// proxy mac from gatway
+	if !currentNodeOwnsSubnet && vmRunningAtCurrentNode {
```
So on a node that does not own the subnet but has the VM migrated to we send GARPs for all IPs that are running on the original node?

qinqon replied:
Right, since VM is at node that need to do routing to access IPs at the VM's subnet we need to update VM's neighbor cache so those IPs access over the gateway since they are not part of the node's switch.

If we don't do so, the traffic will be down until the ip neighors at VM invalidates and send a normal ARP that will be responded with the proxy ARP and traffic will work, but that can take at to 10 seconds or so from what I have see, and depends on guest VM network configuration.

kyrtapz commented on go-controller/pkg/kubevirt/router.go:
```diff
@@ -243,29 +296,38 @@ func EnsureRemoteZonePodAddressesToNodeRoute(watchFactory *factory.WatchFactory,
+		return err
+	}
+	if !vmRunningAtNodeOwningSubnet && currentNodeOwnsSubnet {
+		// If we don't do this ARPs will be answer again with wrong mac
```
Why does that happen? Shouldn't the old pod get removed?

kyrtapz reviewed (commented):
Thanks @qinqon! I have some comments.
Just for my knowledge, live-migration is only supported for l2 and default networks is it?

qinqon commented:
> Thanks @qinqon! I have some comments. Just for my knowledge, live-migration is only supported for l2 and default networks is it?

@kyrtapz yes, we just support default pod network for special annotated VMs (the ones used at hypershift), and normal layer2.

qinqon commented on go-controller/pkg/kubevirt/pod.go:
```diff
@@ -683,3 +711,63 @@ func newRouterAdvertisementWithPrefixInfos(ip net.IP, destinationMAC net.Hardwar
+		return nil, nil
+	}
+
+	pods, err := watchFactory.GetAllPods()
```
@kyrtapz there is no label that denotate the node so we cannot filter the the pods by node from the informer also there is no label to denotate default pod network so we cannot filter either by that.

qinqon commented on go-controller/pkg/kubevirt/switch.go:
```diff
@@ -53,3 +66,58 @@ func ComposeARPProxyLSPOption() string {
+			continue
+		}
+		logicalSwitch := &nbdb.LogicalSwitch{
+			Name: vmPod.Spec.NodeName,
```
What do you mean ? we are removing the LSP from running pods that are not the live migration target pods, that's the whole purpose of this.

qinqon commented on go-controller/pkg/ovn/ovn.go:
```diff
@@ -208,7 +208,7 @@ func (oc *DefaultNetworkController) ensureLocalZonePod(oldPod, pod *corev1.Pod,
 	if kubevirt.IsPodLiveMigratable(pod) {
 		v4Subnets, v6Subnets := util.GetClusterSubnetsWithHostPrefix()
-		return kubevirt.EnsureLocalZonePodAddressesToNodeRoute(oc.watchFactory, oc.nbClient, oc.lsManager, pod, ovntypes.DefaultNetworkName, append(v4Subnets, v6Subnets...))
+		return kubevirt.ReconcileClusterDefaultNetworkLocalZone(oc.watchFactory, oc.nbClient, oc.lsManager, pod, ovntypes.DefaultNetworkName, append(v4Subnets, v6Subnets...))
```
Agree, will fix that, I doubt we will ever implement this for UDN layer3.

qinqon commented on go-controller/pkg/kubevirt/pod.go:
```diff
@@ -237,7 +238,30 @@ func nodeContainsPodSubnet(watchFactory *factory.WatchFactory, nodeName string,
+		}
+	}
+	return nil, nil
+}
```
Make sense I will change it.

kyrtapz commented on go-controller/pkg/kubevirt/router.go:
```diff
@@ -243,29 +296,38 @@ func EnsureRemoteZonePodAddressesToNodeRoute(watchFactory *factory.WatchFactory,
-	// that target VM pod is ready to receive traffic so we can route
-	// taffic to it.
-	targetReadyTimestamp := pod.Annotations[kubevirtv1.MigrationTargetReadyTimestamp]
+	currentNode := os.Getenv("K8S_NODE")
```
This seems to be the first time we introduce direct dependency on `K8S_NODE` in the code, it means with this change all deployment types need to ensure it is set.
I think an alternative might be to depend on the `zone` name but I don't like it all that much either so I am ok to keep this one as is. 
Let's just check it is set, for sanity.

kyrtapz commented on go-controller/pkg/kubevirt/switch.go:
```diff
@@ -53,3 +66,58 @@ func ComposeARPProxyLSPOption() string {
+		}
+		lsp, err := libovsdbops.GetLogicalSwitchPort(nbClient, lsp)
+		if err != nil {
+			continue
```
NIT: Let's log the error?

kyrtapz reviewed (commented):
Looks good to me, ready for a second pair of eyes. 
One concern is that this will be quite heavy when there is a lot of pods and VMs spread out across the cluster, I don't see a better solution though. 
Few minor comments.

qinqon commented:
> Looks good to me, ready for a second pair of eyes. One concern is that this will be quite heavy when there is a lot of pods and VMs spread out across the cluster, I don't see a better solution though. Few minor comments.

Although this is true, the scenario is very niche, since this is done for default pod network VMs with the migration annotation, so more or less hypershift.

I don't know if implementing some kind of parallelization would help here.

qinqon commented:
Something got merged and east/west iperf connectins get broken at default pod network during live migration

qinqon committed:
kubevirt, CDN: Reconcile neighbors after live migration

After VM live migration to a different node, neighbor cache entries (ARP/NDP)
need to be updated to ensure proper traffic routing. This commit adds neighbor
reconciliation logic to update ARP/NDP caches with the arp_proxy MAC address
when VMs migrate between nodes with different subnet ownership.

Since PR#5773 changed DHCP to advertise the actual subnet gateway IP instead
of the link-local 169.254.1.1 address, VMs resolve the gateway to the node's
real LRP MAC on the subnet-owning switch (higher priority than arp_proxy).
After migration, this LRP MAC doesn't exist on the destination switch, so we
must also GARP the gateway IP to update the VM's neighbor entry to the
arp_proxy MAC. This issue was masked in e2e tests by PR#4755 which changed
the connectivity test checks.

The reconciliation handles:
- At migration target (local zone): notify the VM that subnet peers and the
  gateway IP should use the arp_proxy MAC via GARPs/unsolicited NAs
- At subnet-owning node (remote zone): delete stale source pod LSPs to
  prevent wrong ARP responses, and notify remaining pods to use the arp_proxy
  MAC for the migrated VM's IP

References:
- Bug introduced by: https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5773
- Masked in tests by: https://github.com/ovn-kubernetes/ovn-kubernetes/pull/4755

Signed-off-by: Enrique Llorente <ellorent@redhat.com>

qinqon committed:
wip

Signed-off-by: Enrique Llorente <ellorent@redhat.com>
